File: android/src/main/java/com/authsignal/react/AuthsignalModule.java
package com.authsignal.react;

import android.app.Activity;
import android.content.Intent;
import android.content.ActivityNotFoundException;
import android.net.Uri;
import androidx.annotation.NonNull;

import com.facebook.react.bridge.ActivityEventListener;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;

import java.net.URL;
import java.net.URLDecoder;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

import static android.app.Activity.RESULT_OK;

public class AuthsignalModule extends ReactContextBaseJavaModule implements ActivityEventListener {
    private final ReactApplicationContext reactContext;
    private Callback callback;

    public AuthsignalModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
        this.reactContext.addActivityEventListener(this);
    }

    @Override
    public Map<String, Object> getConstants() {
        final Map<String, Object> constants = new HashMap<>();
        constants.put("bundleIdentifier", reactContext.getApplicationInfo().packageName);
        return constants;
    }

    @NonNull
    @Override
    public String getName() {
        return "AuthsignalModule";
    }

    @ReactMethod
    public void launch(String url, Callback callback) {
        final Activity activity = getCurrentActivity();
        final Uri parsedUrl = Uri.parse(url);
        this.callback = callback;

        try {
            if (activity != null) {
                AuthenticationActivity.authenticateUsingBrowser(activity, parsedUrl);
            } else {
                final WritableMap error = Arguments.createMap();
                error.putString("error", "activity_not_available");
                error.putString("error_description", "Android Activity is null.");
                callback.invoke(error);
            }
        } catch (ActivityNotFoundException e){
            final WritableMap error = Arguments.createMap();
            error.putString("error", "browser_not_available");
            error.putString("error_description", "No browser app is installed");
            callback.invoke(error);
        }
    }

    @Override
    public void onActivityResult(Activity activity, int requestCode, int resultCode, Intent data) {
        Callback cb = AuthsignalModule.this.callback;

        if (cb == null) {
            return;
        }

        boolean hasResult = resultCode == RESULT_OK &&
                requestCode == AuthenticationActivity.AUTHENTICATION_REQUEST &&
                data.getData() != null;

        if (hasResult) {
            try {
                String redirectUrl = data.getData().toString();
                String query = redirectUrl.split("[?]")[1];
                String[] pairs = query.split("&");
                String token = null;
                for (String pair : pairs) {
                    int index = pair.indexOf("=");
                    String name = URLDecoder.decode(pair.substring(0, index), "UTF-8");
                    String value = URLDecoder.decode(pair.substring(index + 1), "UTF-8");
                    if (name.equals("token")) {
                        token = value;
                    }
                }
                cb.invoke(null, token);
            } catch (Exception ex) {
                final WritableMap error = Arguments.createMap();
                error.putString("error", "malformed_url");
                error.putString("error_description", "Malformed redirect url");
                cb.invoke(error);
            }

        } else {
            final WritableMap error = Arguments.createMap();
            error.putString("error", "user_cancelled");
            error.putString("error_description", "User cancelled");
            cb.invoke(error);
        }

        AuthsignalModule.this.callback = null;
    }

    @Override
    public void onNewIntent(Intent intent) {
    }
}

File: android/src/main/java/com/authsignal/react/AuthsignalModule.kt
package com.authsignal.react

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Intent
import android.net.Uri
import com.authsignal.DeviceCache
import com.authsignal.TokenCache.Companion.shared
import com.authsignal.react.AuthenticationActivity.Companion.AUTHENTICATION_REQUEST
import com.authsignal.react.AuthenticationActivity.Companion.CALLBACK_SCHEME
import com.authsignal.react.AuthenticationActivity.Companion.EXTRA_ERROR_CODE
import com.authsignal.react.AuthenticationActivity.Companion.EXTRA_ERROR_MESSAGE
import com.authsignal.react.AuthenticationActivity.Companion.authenticateUsingBrowser
import com.facebook.react.bridge.ActivityEventListener
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.module.annotations.ReactModule
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

@ReactModule(name = AuthsignalModule.NAME)
class AuthsignalModule(private val reactContext: ReactApplicationContext) :
  NativeAuthsignalModuleSpec(reactContext), ActivityEventListener {
  private val coroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
  private var launchPromise: Promise? = null

  init {
    reactContext.addActivityEventListener(this)
    DeviceCache.shared.initialize(reactContext.applicationContext)
  }

  @ReactMethod
  override fun setToken(token: String?, promise: Promise) {
    shared.token = token

    promise.resolve("token_set")
  }

  @ReactMethod
  override fun launch(url: String?, promise: Promise) {
    if (launchPromise != null) {
      promise.reject(
        "launch_in_progress",
        "Another Authsignal browser flow is already in progress."
      )
      return
    }

    if (url == null) {
      promise.reject("invalid_url", "Launch URL must not be null.")
      return
    }

    val activity = reactContext.currentActivity
    val parsedUrl = buildLaunchUri(url)
    this.launchPromise = promise

    try {
      if (activity != null) {
        authenticateUsingBrowser(activity, parsedUrl)
      } else {
        promise.reject("activity_not_available", "Android Activity is null.")
        this.launchPromise = null
      }
    } catch (e: ActivityNotFoundException) {
      promise.reject("browser_not_available", "No browser app is installed")
      this.launchPromise = null
    }
  }

  override fun onActivityResult(
    activity: Activity,
    requestCode: Int,
    resultCode: Int,
    data: Intent?
  ) {
    if (requestCode != AUTHENTICATION_REQUEST) return

    val pendingPromise = launchPromise ?: return
    launchPromise = null

    val errorCode = data?.getStringExtra(EXTRA_ERROR_CODE)
    if (errorCode != null) {
      pendingPromise.reject(
        errorCode,
        data.getStringExtra(EXTRA_ERROR_MESSAGE) ?: "Authentication failed."
      )
      return
    }

    if (resultCode == Activity.RESULT_CANCELED) {
      pendingPromise.resolve(null)
      return
    }

    val redirectUri = data?.data
    if (resultCode != Activity.RESULT_OK || redirectUri == null) {
      pendingPromise.reject("malformed_url", "Malformed redirect URL.")
      return
    }

    val token = redirectUri.getQueryParameter("token")
    if (token != null) {
      shared.token = token
    }
    pendingPromise.resolve(token)
  }

  @ReactMethod
  override fun getDeviceId(promise: Promise) {
    coroutineScope.launch {
      val deviceId = DeviceCache.shared.getDefaultDeviceId()
      promise.resolve(deviceId)
    }
  }

  override fun onNewIntent(intent: Intent) {
  }

  private fun buildLaunchUri(url: String): Uri {
    val parsedUrl = Uri.parse(url)
    val builder = parsedUrl.buildUpon().clearQuery()

    for (queryName in parsedUrl.queryParameterNames) {
      if (queryName == NATIVE_SCHEME_QUERY_PARAM) continue

      for (queryValue in parsedUrl.getQueryParameters(queryName)) {
        builder.appendQueryParameter(queryName, queryValue)
      }
    }

    builder.appendQueryParameter(NATIVE_SCHEME_QUERY_PARAM, CALLBACK_SCHEME)

    return builder.build()
  }

  companion object {
    const val NAME = "AuthsignalModule"
    private const val NATIVE_SCHEME_QUERY_PARAM = "nativeScheme"
  }
}
